### ISSUE TITLE
Document setter context

### ISSUE THREAD

**Issue by ethanresnick** [Automattic/mongoose#6209]
As of https://github.com/Automattic/mongoose/issues/5340, it seems that setters run on all queries, which is nice. This will obviously blow up, though, if the setter is expecting `this` to be a document (as I believe it was in Mongoose 4). Therefore, it'd be good to document what `this` is now when setters run against a query. I assume it's the query object? (If it's not the query object, maybe it should be?)

**Comment by lineus** [comment]
Hi @ethanresnick, I'm pretty new to mongoose, so I wanted to verify the context of the setter in mongoose 5. here is a small script that shows that `this` changes as you expected:
```
#!/usr/bin/env node
'use strict'

const mongoose = require('./lib/test_db')
const Schema = mongoose.Schema

function findContext (str) {
  let msg = 'setter context is a query:'
  console.log(msg, this instanceof mongoose.Query)
  return String(str).toUpperCase()
}

const testSchema = new Schema({
  name: {
    type: String,
    set: findContext
  }
})

const Test = mongoose.model('tests', testSchema)

process.stdout.write('Model instance -> ')

const test = new Test({
  name: 'Thornton'
})

test.save().then((res) => {
  console.log(res)
  let con = { _id: test._id }
  let upd = { name: 'Huffy' }
  let opt = { new: true }
  process.stdout.write('Query -> ')
  Test.findOneAndUpdate(con, upd, opt, (err, updated) => {
    if (err) { return console.error(err) }
    mongoose.connection.close()
    return console.log(updated)
  })
}, (rej) => {
  return console.error(rej)
})
```
### Output:
```
InspiredMacPro:mongoose5 lineus$ ./6209.js 
Model instance -> setter context is a query: false // obvious I know.. :)
{ name: 'THORNTON', _id: 5a9e7614fa2d7a49657935be, __v: 0 }
Query -> setter context is a query: true // V E R I F I E D
{ _id: 5a9e7614fa2d7a49657935be, name: 'HUFFY', __v: 0 }
InspiredMacPro:mongoose5 lineus$ 
```
I am going to create a pull request for a new page in the guide, something like which_this.html that contains all of the various places one might have to be aware of this, ie. getters and setters, middleware, etc. I'll include examples.  Unless @vkarpov15 or @varunjayaraman think it's a bad idea.

**Comment by varunjayaraman** [comment]
here's a more complete repro script thats copy-pastable (you forgot to establish a mongoose connection @lineus ):

```javascript
const mongoose = require('mongoose')
const Schema = mongoose.Schema

function findContext(str) {
  let msg = 'setter context is a query:'
  console.log(msg, this instanceof mongoose.Query)
  return String(str).toUpperCase()
}

const testSchema = new Schema({
  name: {
    type: String,
    set: findContext
  }
})

const Test = mongoose.model('tests', testSchema)

process.stdout.write('Model instance -> ')

const test = new Test({
  name: 'Thornton'
})


mongoose
  .connect('mongodb://localhost:27017/gh-6209')
  .then(() => test.save())
  .then((res) => {
    console.log(res)
    let con = { _id: test._id }
    let upd = { name: 'Huffy' }
    let opt = { new: true }
    process.stdout.write('Query -> ')
    return Test.findOneAndUpdate(con, upd, opt, (err, updated) => {
      if (err) { return console.error(err) }
      mongoose.connection.close()
      return console.log(updated)
    })
  })
  .catch(error => console.error(error.stack))
```

I agree it can be confusing determining what this is. I think a page on what `this` is might be unmanageable, but it would be worth it to maybe drop notes for the part on running setters on queries that `this` is indeed now the mongoose query object

**Comment by lineus** [comment]
@varunjayaraman Thanks for the input. The manageability of the this page had crossed my mind too, I figured I would wait until I heard from you before I did anything.

In my example above I import mongoose with
`const mongoose = require('./lib/test_db')`
here's whats in that file:
```
$ cat ./lib/test_db.js 
const uri = 'mongodb://localhost:27017/test'
const mongoose = require('mongoose')

let mVer = mongoose.version
let mongoOpts = {}

if (mVer.match(/^4/)) {
  mongoose.Promise = global.Promise
  mongoOpts.useMongoClient = true
}

mongoose.connect(uri, mongoOpts)

module.exports = mongoose
$ 
```
*tl;dr If you see me put code and output in a bug, you can be sure I actually ran it :)*

Being so new to this, I create a file for every example from issues that I comment on and tinker with them until I think I understand whats going on. I won't do this on every one forever, just until I stop learning something new every time.

In the past when I thought someone was new to this like I am, or certainly if the question is about the connection, I would skip the module above and handle everything explicitly in my example. 

If you think I should skip the module all of the time, I'm totally happy to do that.
